dask/tests/test_highgraph.py

```diff
@@ -130,3 +131,54 @@ def test_blockwise_cull(flat):
         out_keys = layer.get_output_keys()
         assert out_keys == {(layer.output, *select)}
         assert not layer.is_materialized()
+
+
+def test_layer_interface():
+    # Import collections to populate Layer.__subclasses__
+    import dask.array  # noqa
+    import dask.dataframe  # noqa
+    import dask.bag  # noqa
+    import dask.delayed  # noqa
+
+    stack = {Layer}
+    checked = {".".join((Layer.__module__, Layer.__name__))}
+
+    while stack:
+        cls = stack.pop()
+
+        for subclass in cls.__subclasses__():
+            stack.add(subclass)
+            checked.add(".".join((subclass.__module__, subclass.__name__)))
+
+        members = dict(inspect.getmembers(cls))
+        # Must be able to pickle it
+        assert "__reduce__" in members
+        spec = inspect.getfullargspec(members["__reduce__"])
+        assert spec.args == ["self"]
+
+        # Must be able to pack it
+        assert "__dask_distributed_pack__" in members
+        spec = inspect.getfullargspec(members["__dask_distributed_pack__"])
+        assert spec.args == ["self", "client"]
+        pack = getattr(cls, "__dask_distributed_pack__")
+        assert inspect.isfunction(pack) and not hasattr(pack, "__self__")
+
+        # Must be able to unpack it
+        assert "__dask_distributed_unpack__" in members
+        spec = inspect.getfullargspec(members["__dask_distributed_unpack__"])
+        assert spec.args == ["cls", "state", "dsk", "dependencies", "annotations"]
+        unpack = getattr(cls, "__dask_distributed_unpack__")
+        # https://stackoverflow.com/a/19228282
+        assert inspect.ismethod(unpack) and unpack.__self__ is cls
+
+    assert checked == {
+        "dask.dataframe.io.csv.BlockwiseReadCSV",
+        "dask.dataframe.shuffle.SimpleShuffleLayer",
+        "dask.dataframe.io.parquet.core.BlockwiseParquet",
+        "dask.dataframe.shuffle.ShuffleLayer",
+        "dask.blockwise.BlockwiseIO",
+        "dask.dataframe.io.parquet.core.ParquetSubgraph",
+        "dask.highlevelgraph.Layer",
+        "dask.highlevelgraph.BasicLayer",
+        "dask.blockwise.Blockwise",
+    }
```

Comment:
This test seems very brittle. If we change the interface in a sensible way then this will break.  If we add any new classes then this will break.  This adds a burden to development that I think we shouldn't add.

Instead, I think that you're trying to verify that the interface is consistent.  If so, I wonder if you're looking to add Abstract Base Class functionality to `Layer`?  

When I suggested a test I was more thinking of something that a user would run into.  You're running into this somewhere today.  Let's make sure that a simplified version of that workload will continue to work in the future.  I'm guessing that this involves a small computation that involves a BlockwiseIO layer and the distributed scheduler.  There is a place in dask/dask for distributed tests, test_distributed.py I think.  Maybe we want a test like the following:

```python
@gen_cluster(client=True)
def test_annotations_on_io_layers(c, s, a, b):
    with dask.annotate(workers=a.address):
        x = da.ones(10, chunks=(5,))
    x = await x.persist()
    assert a.data and not b.data
```

Reply:
In general, as you're learning, I'm allergic to tests that are highly tied to implementations.  My experience is that tests like these add a lot of inertia to codebases.  They're great for short-term development but less great for long-term and cross-team development.

Reply:
>This test seems very brittle. If we change the interface in a sensible way then this will break. If we add any new classes then this will break. This adds a burden to development that I think we shouldn't add.

> Instead, I think that you're trying to verify that the interface is consistent. If so, I wonder if you're looking to add Abstract Base Class functionality to Layer?

An @abstractmethod would be one way of doing it, but it feels like the Layer hierarchy is in flux at the moment: For example if__dask_distributed_unpack__ was decorated with @abc.abstractmethod, then BasicLayer would be forced to implement it. But it doesn't, instead it's functionality seems to be implemented in distributed via ``distributed.protocol.highlevelgraph._materialized_layer_pack`` and  ``distributed.protocol.highlevelgraph._materialized_layer_unpack``. I don't understand the reason for the split, so my philosophy has been not to tinker with Layer too much.

Thus, the test is indeed brittle, but I'd argue that the interface + hierarchy is a bit brittle at present. I think it's better to explicitly fail and nudge the developer, who is probably fairly sophisticated if they're implementing their own Layer.

Thoughts?

Reply:
> In general, as you're learning, I'm allergic to tests that are highly tied to implementations. My experience is that tests like these add a lot of inertia to codebases. They're great for short-term development but less great for long-term and cross-team development.

I saw this just after you replied. Generally I'm happy to follow the philosophy, but my feeling is that we're having this conversation because the Layer hierarchy isn't finalised. It seems that interface-wise, subclassing a Layer involves:

1. __dask_distributed_pack__
2. __dask_distributed_unpack__
3. __reduce__

and consider annotations in each one. I think my allergy in this case is interfering with an implementation spread across multiple packages -- better just to check that the interface is adhered too?

Reply:
Perhaps the greater question I should be asking is: how does one write a test case that ensures that annotations are transmitted correctly for all layers? As packing/unpacking each Layer involves handling annotations, this may end up requiring an annotation test case for each Layer subclass.

One option may be to separate annotation packing/unpacking functionality out from each Layer implementation (its been the same for all subclasses thus far).